src/modes/ModeEstimation.jl

```diff
@@ -0,0 +1,318 @@
+
+A struct that stores the log density function of a `DynamicPPL` model.
+"""
+struct OptimLogDensity{M<:Model,C<:AbstractContext,V<:VarInfo}
+    "A `DynamicPPL.Model` constructed either with the `@model` macro or manually."
+    model::M
+    "A `DynamicPPL.AbstractContext` used to evaluate the model. `LikelihoodContext` or `DefaultContext` are typical for MAP/MLE."
+    context::C
+    "A `DynamicPPL.VarInfo` struct that will be used to update model parameters."
+    vi::V
+    "Whether the to evaluate the model with transformed variables."
+    linked::Bool
+end
+
+"""
+    OptimLogDensity(model::Model, context::AbstractContext)
+
+Create a callable `OptimLogDensity` struct that evaluates a model using the given `context`.
+"""
+function OptimLogDensity(model::Model, context::AbstractContext, linked::Bool)
+	init = VarInfo(model, context)
+	linked && DynamicPPL.link!(init, DynamicPPL.SampleFromPrior())
+	return OptimLogDensity(model, context, init, linked)
+end
+
+"""
+    (f::OptimLogDensity)(z; unlinked::Bool = false)
+
+Evaluate the log joint (with `DefaultContext`) or log likelihood (with `LikelihoodContext`)
+at the array `z`. If `unlinked=true`, no change of variables will occur.
+"""
+function (f::OptimLogDensity)(z; linked::Bool = true)
+    spl = DynamicPPL.SampleFromPrior()
+
+    varinfo = DynamicPPL.VarInfo(f.vi, spl, z)
+    f.model(varinfo, spl, f.context)
+    return -DynamicPPL.getlogp(varinfo)
+end
+
+"""
+    unlink(f::OptimLogDensity)
+
+Generate an unlinked (with no variable transformions) version of an existing `OptimLogDensity`.
+"""
+function Bijectors.invlink(f::OptimLogDensity)
+    init = VarInfo(f.model, f.context)
+    return OptimLogDensity(f.model, f.context, init, false)
+end
+
+"""
+    link(f::OptimLogDensity)
+
+Generate an linked (with variable transformions) version of an existing `OptimLogDensity`.
+"""
+function Bijectors.link(f::OptimLogDensity)
+    init = VarInfo(f.model, f.context)
+    DynamicPPL.link!(init, DynamicPPL.SampleFromPrior())
+    return OptimLogDensity(f.model, f.context, init, false)
+end
+
+"""
+    ModeResult{
+        V<:NamedArrays.NamedArray, 
+        M<:NamedArrays.NamedArray, 
+        O<:Optim.MultivariateOptimizationResults, 
+        S<:NamedArrays.NamedArray
+    }
+
+A wrapper struct to store various results from a MAP or MLE estimation.
+"""
+struct ModeResult{
+    V<:NamedArrays.NamedArray, 
+    O<:Optim.MultivariateOptimizationResults,
+    M<:OptimLogDensity
+} <: StatsBase.StatisticalModel
+    "A vector with the resulting point estimates."
+    values :: V
+    "The stored Optim.jl results."
+    optim_result :: O
+    "The final log likelihood or log joint, depending on whether `MAP` or `MLE` was run."
+    lp :: Float64
+    "The evaluation function used to calculate the output."
+    f :: M
+end
+#############################
+# Various StatsBase methods #
+#############################
+
+function Base.show(io::IO, m::ModeResult)
```

Comment:
Maybe change this to
```suggestion
function Base.show(io::IO, ::MIME"text/plain", m::ModeResult)
```
and implement another more compact `show` method with just one line (that would be used, e.g., when displaying arrays of ModeResult)? See https://docs.julialang.org/en/v1/manual/types/#man-custom-pretty-printing-1.